Userland/Libraries/LibAudio/ConnectionFromClient.cpp

```diff
@@ -1,48 +1,126 @@
 /*
  * Copyright (c) 2018-2020, Andreas Kling <kling@serenityos.org>
+ * Copyright (c) 2022, kleines Filmröllchen <filmroellchen@serenityos.org>
  *
  * SPDX-License-Identifier: BSD-2-Clause
  */
 
-#include <LibAudio/Buffer.h>
+#include <AK/Atomic.h>
+#include <AK/Format.h>
+#include <AK/OwnPtr.h>
+#include <AK/Time.h>
+#include <AK/Types.h>
 #include <LibAudio/ConnectionFromClient.h>
+#include <LibAudio/UserSampleQueue.h>
+#include <LibCore/Event.h>
 #include <time.h>
 
 namespace Audio {
 
-// FIXME: We don't know what is a good value for this.
-// Real-time audio may be improved with a lower value.
-static timespec g_enqueue_wait_time { 0, 10'000'000 };
-
 ConnectionFromClient::ConnectionFromClient(NonnullOwnPtr<Core::Stream::LocalSocket> socket)
     : IPC::ConnectionToServer<AudioClientEndpoint, AudioServerEndpoint>(*this, move(socket))
+    , m_buffer(make<AudioQueue>(MUST(AudioQueue::try_create())))
+    , m_user_queue(make<UserSampleQueue>())
+    , m_background_audio_enqueuer(Threading::Thread::construct([this]() {
+        // All the background thread does is run an event loop
+        Core::EventLoop enqueuer_loop;
+        m_enqueuer_loop = &enqueuer_loop;
+        enqueuer_loop.exec();
+        m_enqueuer_loop = nullptr;
+        return (intptr_t) nullptr;
+    }))
+{
+    m_background_audio_enqueuer->start();
+    set_buffer(*m_buffer);
+}
+
+ConnectionFromClient::~ConnectionFromClient()
 {
+    die();
 }
 
-void ConnectionFromClient::enqueue(Buffer const& buffer)
+void ConnectionFromClient::die()
 {
-    for (;;) {
-        auto success = enqueue_buffer(buffer.anonymous_buffer(), buffer.id(), buffer.sample_count());
-        if (success)
+    // We're sometimes getting here after the other thread has already exited and its event loop does no longer exist.
+    if (m_enqueuer_loop != nullptr) {
```

Comment:
This check for `m_enqueuer_loop` is racy, as it happens on the background thread.
That could lead us to corrupting memory by calling methods on a stack object, for a thread that is no longer present.